### DISCUSSION TITLE
Add a new repos for goodies?

### DISCUSSION THREAD

**Discussion by PMassicotte** [R-nvim/R.nvim#66] [Ideas]
Over the time I have made some function that might be nice to add but are not required for R.nvim to work.

For example, I have a helper function that will format SQL queries:

![Peek 2024-03-03 11-55](https://github.com/R-nvim/R.nvim/assets/4519221/cb5c6a0b-58f8-45b6-aa56-27813498db72)

Do you think such thing could be part of an optional plugin?

**Comment by jalvesaq** [comment]
Yes, I think that features that few people are going to use can be included in other plugins in the organization.

Is the screenshot from an R file? If yes, how did the SQL string get highlighted?

**Comment by PMassicotte** [comment]
yes from an R file. I am using treesitter injection for syntax coloration :)

**Comment by PMassicotte** [comment]
Maybe the syntax highlight could be added as an optional option in the config.

**Comment by jalvesaq** [comment]
I don't run SQL from R, but I know it's important for people working with big data. Both the highlighting and the formatting are very nice. Perhaps you could include both in `R.nvim`. I can't do it because I don't use SQL from R yet. But, possibly, in the near future, I will have to use either SQL or Python to manipulate big data.

**Comment by PMassicotte** [comment]
Ok cool I will work on that.

I have started to use duckdb in R a lot recently, and it is fantastic!

**Comment by jalvesaq** [comment]
I see general support for SQL and Python integration with R is appropriate for `R.nvim`. We may also improve support for `tidyverse` packages because they have become mainstream. Additional support for specific packages might be better added in separate plugins because it's hard to stop maintaining something after including it. Some features introduced in `Vim-R-plugin` more than ten years ago are less useful today, but we can't remove them because some people are used to them. So, if an R package is superseded by another one, `R.nvim` will not keep carrying less useful features.

**Comment by PMassicotte** [comment]
I agree. It would be easier to stop maintaining deprecated package. How do you think support for `tidyverse` could be better or different from another package?

**Comment by jalvesaq** [comment]
The support for completing columns of piped `data.frame`s was added because this was becoming increasingly common after `tidyverse` was widely adopted. And perhaps the support can still improve. For example, we don't get completion for the `mutate()` arguments in the code below:

```r
library(tidyverse)
mtcars |>
    filter(hp < 120 &
           mpg > 20) |>
    mutate(kml = mpg / 2.3521458)
```

I'm not sure if `R.nvim` should complete `mpg` in the `mutate()` function because the original `mtcars` is no longer the `.data` argument and the already filtered `data.frame` might even be empty.

**Comment by ltuijnder** [comment]
While I agree that `tidyverse` is popular, I would not call it mainstream. `data.table` is also a very popular choice duo being a fast and lightweight alternative to `dplyr`. On top of that `data.table` also comes with its own non-standard evaluation syntax:
```
library(data.table)
dt = data.table(mtcars)
result = dt[hp<120 & mpg > 20][, kml := mpg/2.3421458]
```
I do not think it is a good idea for `R.nvim` to be opinionated in which table manipulation library it supports out of the box and leave other packages out. If you are going to support tidyverse you will probably also get feature request for other packages (like `data.table`). 
Personally, I use `data.table` a lot and not so much tidyverse duo its gigantic dependency footprint which I as a package author am conscious about.

That said, smart autocompletion about columns of the table for popular table package would be nice. :) But, like already mentioned, this can be hard to get right since you need to evaluate the code to correctly know which columns are still present in an evaluation. Which for large datasets would make the editor slow to respond. Or maybe I am missing here some technical implementation that that could get this to work efficiently.

**Reply by ssh352** [reply]
does Rstudio or vscode complete on data.table columns?

**Reply by ltuijnder** [reply]
sorry for the late reply. I do not have experience with coding R in Rstudio or vscode. I used to code in Architect. (an eclipse-based IDE. And there it did not complete in)

Now I use radian and rlanguageserver in nvim and there I also do not have autocompletion on the columns. 

VScode autocompletion is also based on the rlanguagueserver: https://code.visualstudio.com/docs/languages/r#_code-completion-intellisense 
R studio I do not know. But I can imagine that they maybe did the effort to support their dplyr tidyverse completion. But frankly, I am not sure.

**Comment by PMassicotte** [comment]
Any interest in this:

![Peek 2024-06-21 14-14](https://github.com/R-nvim/R.nvim/assets/4519221/842f9f62-e7b8-4c9e-98fd-b96ce928b9cf)

**Comment by jalvesaq** [comment]
Yes, it's interesting! It's the kind of feature we expect to find in a language server.

**Reply by PMassicotte** [reply]
good, should be able to make a PR in the next few days!
